### ISSUE TITLE
Linking error on Debian testing

### ISSUE THREAD

**Issue by sylvestre** [tensorflow/mlir#16]
Linking error 
```
cd "/home/sylvestre/dev/debian/pkg-llvm/llvm-toolchain/branches/llvm-toolchain-snapshot_9~svn357703/build-llvm/projects/mlir/tools/mlir-cpu-runner" && /usr/bin/cmake -E cmake_link_script CMakeFiles/mlir-cpu-runner.dir/link.txt --verbose=1
/usr/bin/g++-8  -g -O2 -fdebug-prefix-map=/home/sylvestre/dev/debian/pkg-llvm/llvm-toolchain/branches/llvm-toolchain-snapshot_9~svn357703=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -ffunction-sections -fdata-sections -O2 -DNDEBUG -g1  -Wl,-z,relro -Wl,-allow-shlib-undefined -Llib -Wl,--whole-archive,-lMLIRLLVMIR,-lMLIRStandardOps,-lMLIRTargetLLVMIR,-lMLIRTransforms,-lMLIRTranslation,--no-whole-archive CMakeFiles/mlir-cpu-runner.dir/mlir-cpu-runner.cpp.o  -o ../../../../bin/mlir-cpu-runner ../../../../lib/libMLIRIR.a ../../../../lib/libMLIRAffineOps.a ../../../../lib/libMLIRAnalysis.a ../../../../lib/libMLIREDSC.a ../../../../lib/libMLIRExecutionEngine.a ../../../../lib/libMLIRIR.a ../../../../lib/libMLIRParser.a ../../../../lib/libMLIRTransforms.a ../../../../lib/libMLIRSupport.a ../../../../lib/libLLVMCore.a ../../../../lib/libLLVMSupport.a ../../../../lib/libMLIRVectorOps.a ../../../../lib/libMLIRLLVMIR.a ../../../../lib/libMLIRPass.a ../../../../lib/libLLVMOrcJIT.a ../../../../lib/libLLVMExecutionEngine.a ../../../../lib/libLLVMRuntimeDyld.a ../../../../lib/libLLVMX86CodeGen.a ../../../../lib/libLLVMAsmPrinter.a ../../../../lib/libLLVMDebugInfoDWARF.a ../../../../lib/libLLVMGlobalISel.a ../../../../lib/libLLVMSelectionDAG.a ../../../../lib/libLLVMCodeGen.a ../../../../lib/libLLVMTarget.a ../../../../lib/libLLVMX86Desc.a ../../../../lib/libLLVMX86AsmPrinter.a ../../../../lib/libLLVMX86Utils.a ../../../../lib/libLLVMMCDisassembler.a ../../../../lib/libLLVMX86Info.a ../../../../lib/libLLVMipo.a ../../../../lib/libLLVMBitWriter.a ../../../../lib/libLLVMIRReader.a ../../../../lib/libLLVMAsmParser.a ../../../../lib/libLLVMInstrumentation.a ../../../../lib/libLLVMLinker.a ../../../../lib/libLLVMScalarOpts.a ../../../../lib/libLLVMAggressiveInstCombine.a ../../../../lib/libLLVMInstCombine.a ../../../../lib/libLLVMVectorize.a ../../../../lib/libLLVMTransformUtils.a ../../../../lib/libLLVMAnalysis.a ../../../../lib/libLLVMObject.a ../../../../lib/libLLVMBitReader.a ../../../../lib/libLLVMMCParser.a ../../../../lib/libLLVMMC.a ../../../../lib/libLLVMDebugInfoCodeView.a ../../../../lib/libLLVMDebugInfoMSF.a ../../../../lib/libLLVMProfileData.a ../../../../lib/libLLVMCore.a ../../../../lib/libLLVMBinaryFormat.a ../../../../lib/libMLIRAnalysis.a ../../../../lib/libMLIRAffineOps.a ../../../../lib/libMLIRIR.a ../../../../lib/libMLIRStandardOps.a ../../../../lib/libLLVMSupport.a -lz -lrt -ldl -ltinfo -lpthread -lm ../../../../lib/libLLVMDemangle.a 
/usr/bin/ld : ne peut trouver -lMLIRLLVMIR
/usr/bin/ld : ne peut trouver -lMLIRStandardOps
/usr/bin/ld : ne peut trouver -lMLIRTargetLLVMIR
/usr/bin/ld : ne peut trouver -lMLIRTransforms
/usr/bin/ld : ne peut trouver -lMLIRTranslation
collect2: error: ld returned 1 exit status
```

Built with:

```
cmake \
-DCMAKE_INSTALL_PREFIX=/usr/lib/llvm-9 \
-DCMAKE_VERBOSE_MAKEFILE=ON \
-DCMAKE_BUILD_TYPE=RelWithDebInfo \
-DCMAKE_CXX_FLAGS_RELWITHDEBINFO="-O2 -DNDEBUG -g1" \
-DBOOTSTRAP_CMAKE_CXX_FLAGS=' -fuse-ld=gold -fPIC -Wno-unused-command-line-argument -Wno-unknown-warning-option ' \
-DBOOTSTRAP_CMAKE_C_FLAGS=' -fuse-ld=gold -fPIC -Wno-unused-command-line-argument -Wno-unknown-warning-option ' \
-DLLVM_LINK_LLVM_DYLIB=ON \
-DLLVM_INSTALL_UTILS=ON \
-DLLVM_VERSION_SUFFIX= \
-DLLVM_ENABLE_SPHINX=ON \
-DSPHINX_WARNINGS_AS_ERRORS=OFF \
-DLLVM_BUILD_LLVM_DYLIB=ON \
-DLLVM_ENABLE_RTTI=ON \
-DLLVM_ENABLE_FFI=ON \
 -DLLVM_BINUTILS_INCDIR=/usr/include/ -DLLVM_USE_PERF=yes \
-DLIBCLANG_LIBRARY_VERSION=1 \
-DENABLE_LINKER_BUILD_ID=ON \
-DPOLLY_BUNDLED_JSONCPP=OFF \
-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="WebAssembly;AVR" \
-DCLANG_ENABLE_BOOTSTRAP=ON \
-DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=ON \
-DCLANG_VENDOR=Debian \
-DCLANG_BOOTSTRAP_PASSTHROUGH="CMAKE_INSTALL_PREFIX;CMAKE_VERBOSE_MAKEFILE;CMAKE_BUILD_TYPE;CMAKE_CXX_FLAGS_RELWITHDEBINFO;LLVM_LINK_LLVM_DYLIB;LLVM_INSTALL_UTILS;LLVM_VERSION_SUFFIX;LLVM_ENABLE_SPHINX;SPHINX_WARNINGS_AS_ERRORS;LLVM_BUILD_LLVM_DYLIB;LLVM_ENABLE_RTTI;LLVM_ENABLE_FFI;LIBCLANG_LIBRARY_VERSION;ENABLE_LINKER_BUILD_ID;POLLY_BUNDLED_JSONCPP;LLVM_EXPERIMENTAL_TARGETS_TO_BUILD;LLVM_USE_PERF;LLVM_ENABLE_ASSERTIONS;LLVM_BINUTILS_INCDIR;LLVM_HOST_TRIPLE;;LLVM_COMPILER_CHECKED;COMPILER_RT_BUILD_BUILTINS;LIBOMP_LIBFLAGS;CMAKE_SHARED_LINKER_FLAGS"

```

**Comment by joker-eph** [comment]
Thanks for the report! We'll look into this after the EuroLLVM conference next week.

**Comment by westion717** [comment]
It's because CMakeLists.txt is using hardcode -L path. 

```
function(whole_archive_link target)
  if("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
    set(link_flags "-Llib -Wl,-all_load ")
    FOREACH(LIB ${ARGN})
      string(CONCAT link_flags ${link_flags} "-l${LIB} ")
    ENDFOREACH(LIB)
  else()
    set(link_flags "**-Llib** -Wl,--whole-archive,")
    FOREACH(LIB ${ARGN})
      string(CONCAT link_flags ${link_flags} "-l${LIB},")
    ENDFOREACH(LIB)
    string(CONCAT link_flags ${link_flags} "--no-whole-archive")
  endif()
  set_target_properties(${target} PROPERTIES LINK_FLAGS ${link_flags})
endfunction(whole_archive_link)
```

**-Llib** I think can be replaced with **-L${LLVM_BINARY_DIR}/lib**

**Comment by jpienaar** [comment]
Thanks, this should have been addressed by commit b65972c.
